File: app/api/ai.ts
import { z } from "zod";
import { createRouter, publicQuery } from "./middleware";
import {
  chatCompletion,
  generateStrategy,
  analyzeMarket,
  listModels,
  healthCheck,
  type OllamaMessage,
} from "./lib/ollama";
import { getDb } from "./queries/connection";
import { chatMessages, researchSessions, strategyInsights } from "../db/schema";
import { eq, desc } from "drizzle-orm";

export const aiRouter = createRouter({
  // Health check for Ollama
  health: publicQuery.query(async () => {
    const ok = await healthCheck();
    const models = ok ? await listModels() : [];
    return { ok, models, message: ok ? "Ollama connected" : "Ollama unavailable - check OLLAMA_URL" };
  }),

  // Chat with AI assistant
  chat: publicQuery
    .input(
      z.object({
        messages: z.array(
          z.object({
            role: z.enum(["system", "user", "assistant"]),
            content: z.string(),
          })
        ),
        model: z.string().optional().default("llama3.1"),
        temperature: z.number().optional().default(0.7),
        sessionId: z.string().optional(),
      })
    )
    .mutation(async ({ input }) => {
      const response = await chatCompletion(
        input.messages as OllamaMessage[],
        input.model,
        { temperature: input.temperature }
      );

      // Save to DB if sessionId provided
      if (input.sessionId) {
        const db = getDb();
        await db.insert(chatMessages).values({
          sessionId: input.sessionId,
          role: "user",
          content: input.messages[input.messages.length - 1].content,
          model: input.model,
        });
        await db.insert(chatMessages).values({
          sessionId: input.sessionId,
          role: "assistant",
          content: response.message.content,
          model: input.model,
        });
      }

      return {
        content: response.message.content,
        model: input.model,
        timing: {
          total: response.total_duration,
          promptTokens: response.prompt_eval_count,
          completionTokens: response.eval_count,
        },
      };
    }),

  // Get chat history for a session
  history: publicQuery
    .input(z.object({ sessionId: z.string() }))
    .query(async ({ input }) => {
      const db = getDb();
      const messages = await db
        .select()
        .from(chatMessages)
        .where(eq(chatMessages.sessionId, input.sessionId))
        .orderBy(chatMessages.createdAt);
      return messages;
    }),

  // Generate strategy analysis with AI
  generateStrategy: publicQuery
    .input(
      z.object({
        symbol: z.string(),
        rr: z.string(),
        features: z.array(z.string()),
        metrics: z.record(z.number()),
        model: z.string().optional().default("llama3.1"),
        sessionId: z.string().optional(),
      })
    )
    .mutation(async ({ input }) => {
      const analysis = await generateStrategy(
        input.symbol,
        input.rr,
        input.features,
        input.metrics,
        input.model
      );

      if (input.sessionId) {
        const db = getDb();
        await db.insert(strategyInsights).values({
          sessionId: input.sessionId,
          strategyName: `${input.symbol}_${input.rr}`,
          insightType: "analysis",
          content: analysis,
          model: input.model,
        });
      }

      return { analysis };
    }),

  // AI market analysis
  marketAnalysis: publicQuery
    .input(
      z.object({
        symbol: z.string(),
        priceData: z.object({
          currentPrice: z.number(),
          change24h: z.number(),
          high24h: z.number(),
          low24h: z.number(),
          volume24h: z.number(),
        }),
        model: z.string().optional().default("llama3.1"),
      })
    )
    .mutation(async ({ input }) => {
      const analysis = await analyzeMarket(
        input.symbol,
        input.priceData,
        input.model
      );
      return { analysis };
    }),

  // Save research session
  saveSession: publicQuery
    .input(
      z.object({
        sessionId: z.string(),
        symbol: z.string(),
        rrConfig: z.string(),
        leverage: z.number().optional().default(10),
        days: z.number().optional().default(60),
        resultJson: z.string().optional(),
        bestStrategy: z.string().optional(),
        winRate: z.number().optional(),
        profitFactor: z.number().optional(),
        expectancy: z.number().optional(),
      })
    )
    .mutation(async ({ input }) => {
      const db = getDb();
      await db.insert(researchSessions).values({
        ...input,
        status: "completed",
      });
      return { success: true };
    }),

  // Get research sessions
  sessions: publicQuery.query(async () => {
    const db = getDb();
    const sessions = await db
      .select()
      .from(researchSessions)
      .orderBy(desc(researchSessions.createdAt))
      .limit(50);
    return sessions;
  }),

  // Get strategy insights
  insights: publicQuery
    .input(z.object({ sessionId: z.string() }))
    .query(async ({ input }) => {
      const db = getDb();
      const insights = await db
        .select()
        .from(strategyInsights)
        .where(eq(strategyInsights.sessionId, input.sessionId))
        .orderBy(desc(strategyInsights.createdAt));
      return insights;
    }),

  // Clear chat history
  clearHistory: publicQuery
    .input(z.object({ sessionId: z.string() }))
    .mutation(async ({ input }) => {
      const db = getDb();
      await db
        .delete(chatMessages)
        .where(eq(chatMessages.sessionId, input.sessionId));
      return { success: true };
    }),
});

File: src/strategy_api/routers/ai.py
from __future__ import annotations
from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, delete, desc
from sqlalchemy.orm import Session
from strategy_api.db.session import get_session
from strategy_api.db.models import ChatMessage, ResearchSession, StrategyInsight
from strategy_api.services import ollama
from strategy_api.schemas import (
    ChatReq,
    GenStrategyReq,
    MarketAnalysisReq,
    SaveSessionReq,
)

router = APIRouter(prefix="/api/ai", tags=["ai"])


@router.get("/health")
async def health():
    ok = await ollama.health_check()
    models = await ollama.list_models() if ok else []
    return {
        "ok": ok,
        "models": models,
        "message": "Ollama connected" if ok else "Ollama unavailable - check OLLAMA_BASE_URL",
    }


@router.post("/chat")
async def chat(req: ChatReq, db: Session = Depends(get_session)):
    res = await ollama.chat_completion(
        [m.model_dump() for m in req.messages], req.model, req.temperature
    )
    content = res["message"]["content"]
    if req.sessionId:
        db.add(ChatMessage(session_id=req.sessionId, role="user",
                           content=req.messages[-1].content, model=req.model))
        db.add(ChatMessage(session_id=req.sessionId, role="assistant",
                           content=content, model=req.model))
        db.commit()
    return {
        "content": content,
        "model": req.model,
        "timing": {
            "total": res.get("total_duration"),
            "promptTokens": res.get("prompt_eval_count"),
            "completionTokens": res.get("eval_count"),
        },
    }


@router.get("/history")
def history(sessionId: str = Query(...), db: Session = Depends(get_session)):
    rows = db.scalars(
        select(ChatMessage).where(ChatMessage.session_id == sessionId)
        .order_by(ChatMessage.created_at)
    ).all()
    return [{"id": r.id, "sessionId": r.session_id, "role": r.role, "content": r.content,
             "model": r.model, "createdAt": r.created_at} for r in rows]


@router.post("/generate-strategy")
async def generate_strategy(req: GenStrategyReq, db: Session = Depends(get_session)):
    analysis = await ollama.generate_strategy(
        req.symbol, req.rr, req.features, req.metrics, req.model
    )
    if req.sessionId:
        db.add(StrategyInsight(session_id=req.sessionId,
                               strategy_name=f"{req.symbol}_{req.rr}",
                               insight_type="analysis", content=analysis, model=req.model))
        db.commit()
    return {"analysis": analysis}


@router.post("/market-analysis")
async def market_analysis(req: MarketAnalysisReq):
    analysis = await ollama.analyze_market(req.symbol, req.priceData.model_dump(), req.model)
    return {"analysis": analysis}


@router.post("/save-session")
def save_session(req: SaveSessionReq, db: Session = Depends(get_session)):
    db.add(ResearchSession(session_id=req.sessionId, symbol=req.symbol, rr_config=req.rrConfig,
                           leverage=req.leverage, days=req.days, status="completed",
                           result_json=req.resultJson, best_strategy=req.bestStrategy,
                           win_rate=req.winRate, profit_factor=req.profitFactor,
                           expectancy=req.expectancy))
    db.commit()
    return {"success": True}


@router.get("/sessions")
def sessions(db: Session = Depends(get_session)):
    rows = db.scalars(
        select(ResearchSession).order_by(desc(ResearchSession.created_at)).limit(50)
    ).all()
    return [{"id": r.id, "sessionId": r.session_id, "symbol": r.symbol, "rrConfig": r.rr_config,
             "leverage": r.leverage, "days": r.days, "status": r.status,
             "bestStrategy": r.best_strategy, "winRate": r.win_rate,
             "profitFactor": r.profit_factor, "expectancy": r.expectancy,
             "createdAt": r.created_at} for r in rows]


@router.get("/insights")
def insights(sessionId: str = Query(...), db: Session = Depends(get_session)):
    rows = db.scalars(
        select(StrategyInsight).where(StrategyInsight.session_id == sessionId)
        .order_by(desc(StrategyInsight.created_at))
    ).all()
    return [{"id": r.id, "sessionId": r.session_id, "strategyName": r.strategy_name,
             "insightType": r.insight_type, "content": r.content, "model": r.model,
             "createdAt": r.created_at} for r in rows]


@router.post("/clear-history")
def clear_history(sessionId: str = Query(...), db: Session = Depends(get_session)):
    db.execute(delete(ChatMessage).where(ChatMessage.session_id == sessionId))
    db.commit()
    return {"success": True}
